lib/libcgroup-3.1.0/doc/man/cgxset.1
.TH CGXSET  1 2022-08-03 "Linux" "libcgroup Manual"
.SH NAME

cgxset \- set the parameters of given cgroup(s)

.SH SYNOPSIS
\fBcgxset\fR [\fB-1\fR] [\fB-2\fR] [\fB-i\fR] [\fB-b\fR] [\fB-r\fR <\fIname=value\fR>] <\fBcgroup_path\fR> ...
.br
\fBcgxset\fR [\fB-b\fR] \fB--copy-from\fR <\fIsource_cgroup_path\fR> <\fBcgroup_path\fR> ...

.SH DESCRIPTION
Set the parameters of input cgroups.
When the request's cgroup version differs from the system's cgroup version, \fBcgxset\fR will
attempt to translate from one version to the other.
If the request version and the system's version are the same, no translation is done and
\fBcgxset\fR will behave like \fBcgset\fR.
If no version is specified by the user (either [\fB-1\fR] or [\fB-2\fR]) and the [\fB-r\fR] flag
is specified, the behavior is undefined.
\fB--copy-from\fR does not require the cgroup version to be specified.

.TP
.B <path>
is the name of the cgroup which should be changed.
This parameter can be used multiple times.

.TP
.B -1, --v1
data in/out of \fBcgxset\fR is in cgroup v1 format.
In this mode, the user will provide values in cgroup v1 format and receive values in v1 format.
format.
If the system is running in cgroup v2 mode, libcgroup will convert the data as necessary

.TP
.B -2, --v2
data in/out of \fBcgxset\fR is in cgroup v2 format.
In this mode, the user will provide values in cgroup v2 format and receive values in v2 format.
If the system is running in cgroup v1 mode, libcgroup will convert the data as necessary

.TP
.B -b
ignores the default systemd delegated hierarchy path and constructs the path of the control groups
relative to the cgroup root hierarchy.

.TP
.B -i, --ignore-unmappable
ignore errors for values that cannot be converted from v1 to v2 or vice versa

.TP
.B -r <name=value>
defines the name of the file to set and
the value which should be written to that file.
This parameter can be used multiple times.

.TP
.B --copy-from <source_cgroup_path>
defines the name of the cgroup whose parameters will be
copied to the input cgroup.

.SH ENVIRONMENT VARIABLES
.TP
.B CGROUP_LOGLEVEL
controls verbosity of the tool. Allowed values are \fBDEBUG\fR,
\fBINFO\fR, \fBWARNING\fR or \fBERROR\fR.

.SH EXAMPLES
.TP
.B cgxset -1 -r cpuset.cpus=0-1 student
set variable cpus in control group student (controller cpuset) to 0-1

.TP
.B cgxset -2 -r cpu.weight=42 first
set variable weight in control group first (controller cpu) to 42

.TP
.B cgxset --copy-from group1/ group2/
copy all parameters of group group1 to group group2
(for all path where both cgroups are defined)


.SH SEE ALSO
cgrules.conf (1), cgcreate (1), cgget (1), cgset (1), cgxget (1)


ChangeLog
DOMjudge Programming Contest Judging System

Version 8.4.0DEV
---------------------------

Version 8.3.0 - 31 May 2024
---------------------------
 - [security] Close metadata file descriptor for the child in runguard.
 - Document that minimum PHP version is now 8.1.0.
 - Upgrade Symfony to 6.2 and upgrade other library dependencies.
 - Show link to problem text on jury problem overview and contest detail pages.
 - Add label to teams to be in sync with ICPC CCS spec.
 - Add feedback/teammessage.txt for automated teamfeedback, see: #2016.
 - Clarifications can now use markdown syntax.
 - Upgrade Bootstrap to 5.2 and dependent components.
 - Bug fixed to correctly select current user during configuration of installation.
 - Properly fix removal of problem/submission data.
 - Filter away balloons of jury members in balloon runner interface.
 - Fix upgrade path for executables, see: #1826.
 - Collect compiler and runner versions before compilation, useful to compare
   judgehosts.
 - Various improvements to shadow config used at ICPC World Finals.
 - Show the historic timelimit which was used when judging the submission.
 - Fixes to the documentation and import of problems with @EXPECTED_RESULTS@
   tags.
 - Allow filtering on affilition in submission page for jury for multisite
   contests.
 - Add preview of the contest problemset to the jury page before contest
   start.
 - Add optional Adminer for easy database access.
 - Fix medal awards when skipping categories.
 - Add direct button to public page for registering when enabled.
 - Scale batch size of judge tasks dynamically.
 - Rename room to location for teams.
 - Add option to upload contest problemset document.
 - Improve on the advice for rejuding on config changes, only when relevant
   advice to start a rejudging, similar for updating the scoreboard cache.
 - For interactive problems: add explicit statement to when one of sides of
   the interactions ran closed their output.
 - Allow importing a (generic) contest warning which is displayed on the scoreboard.
 - Added option in the dj_setup_database to make dumps, useful before major
   upgrades and during the contest before problem imports.
 - Fixed some bugs to become OpenAPI spec compliant again.
 - Prevent a race condition leading to two submissions receiving the same
   colour balloon.
 - Compile executables with `-std=gnu++20` now.
 - Record the testcase directory for judging runs, this makes investigating
   easier.
 - Set default C/C++ standards in configure script, we intend to ship all
   scripts free of warnings/errors.
 - Respect basic authentication in the URL for the import-contest when .netrc
   exists.
 - Fix CORS header for API to make extending with other tools easier.
 - Fix eventfeed when SQL commits happen out of order to send all events.
 - Lazy load the team pictures for the scoreboard, to prevent spamming access
   log with HTTP404s
 - Make TLE detection more robust for interactive problems.
 - Added option for ICPC online scoreboard with no rank displays.

Version 8.2.0 - 6 March 2023
----------------------------
 - Various bugfixes & UI improvements.
 - Add support for Rust as submission language.
 - Add teamphotos to scoreboard zip.
 - Add jury option to to download ZIP with all problem samples, attachments
   and problem statement.
 - Extend team import with optional room and other API extensions, see the API
   doc for further details.
 - Drop support for shadowing a single contest from multiple sources.
 - Log account logins via API separately.
 - Add support for PHP v8.2.
 - Upgraded library dependencies.
 - Removed output of detected entry point to teams.
 - Added option to add custom apt repositories to judge chroot.
 - Added CLI script to configure the domserver with a yaml file.
 - Changed activation behaviour of contests imported via UI to be directly
   available.
 - New option added to reset login status of teams between contests.
 - Highlight submissions which were judged with compiler flags which since
   have changed. Useful when you find wrong flags during the contest.
 - Handle network errors for judgedaemon when domserver is unavailable.
 - Added super lazy mode useful for ICPC analyst instances. Here nothing is
   judged unless manually requested.
 - Split problem import messages in info/warning/danger, to make sure errors
   are not missed when importing.
 - Store user login sessions in database.
 - Contest import CLI script is now rewritten in python.
 - The example test problems are now imported via the contest package format.

Version 8.1.0 - 1 June 2022
---------------------------
 - Document that minimum PHP version is now 7.4.0.
 - Add support for PHP 8.1.
 - Upgrade Symfony to 5.4 LTS and upgrade other library dependencies.
 - Replace calls to deprecated `strftime` with `date`. Note: if you have
   configured a custom `time_format`, you need to update it to the format
   supported by `date`.
 - Dropped the presentation-error verdict, databases from before DOMjudge 5.0
   will be altered to replace the verdict with wrong-answer.
 - Changed API import to give admin users a team.
 - Balloon users can now be created via the API.
 - Shadow mode will now fix missing data by reading from the primary CCS.
 - Fixed cgroup naming for docker to prevent sharing the same container.
 - Removed python2 in the chroot, remove python3 in favor of pypy3 in chroot.
 - Added support for sentry for logging.
 - Added default external IDs on all places.
 - Fixed medals on scoreboard for filtered scoreboards.
 - Dropped old balloons CLI.
 - Reimplementation of runpipe for interactive problems.
 - Individual files can now be removed from executables.
 - Added mysql socket option, for when port/IP is not used.
 - Added support for accounts.{yaml,json} for import script.
 - Added support for judgeremaining for problems.
 - Add option to delete existing testcases/attachments/other data when
   importing a problem.

lib/libcgroup-3.1.0/src/tools/cgxset.c
// SPDX-License-Identifier: LGPL-2.1-only
/**
 * Libcgroup extended cgxset.  Works with both cgroup v1 and v2
 *
 * Copyright (c) 2021-2022 Oracle and/or its affiliates.
 * Author: Tom Hromatka <tom.hromatka@oracle.com>
 */
#include "tools-common.h"
#include "abstraction-common.h"

#include <libcgroup.h>
#include <libcgroup-internal.h>

#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <getopt.h>
#include <stdio.h>

#define FL_RULES	1
#define FL_COPY		2

enum {
	COPY_FROM_OPTION = CHAR_MAX + 1
};

#ifndef UNIT_TEST
static const struct option long_options[] = {
	{"v1",			      no_argument, NULL, '1'},
	{"v2",			      no_argument, NULL, '2'},
	{"ignore-unmappable",	      no_argument, NULL, 'i'},
	{"rule",		required_argument, NULL, 'r'},
	{"help",		      no_argument, NULL, 'h'},
	{"copy-from",		required_argument, NULL, COPY_FROM_OPTION},
	{NULL, 0, NULL, 0}
};

int flags; /* used input method */

static struct cgroup *copy_name_value_from_cgroup(char src_cg_path[FILENAME_MAX])
{
	struct cgroup *src_cgroup;
	int ret = 0;

	/* create source cgroup */
	src_cgroup = cgroup_new_cgroup(src_cg_path);
	if (!src_cgroup) {
		err("can't create cgroup: %s\n", cgroup_strerror(ECGFAIL));
		goto scgroup_err;
	}

	/* copy the name-version values to the cgroup structure */
	ret = cgroup_get_cgroup(src_cgroup);
	if (ret != 0) {
		err("cgroup %s error: %s\n", src_cg_path, cgroup_strerror(ret));
		goto scgroup_err;
	}

	return src_cgroup;

scgroup_err:
	cgroup_free(&src_cgroup);

	return NULL;
}

static void usage(int status, const char *program_name)
{
	if (status != 0) {
		err("Wrong input parameters, ");
		err("try %s --help' for more information.\n", program_name);
		return;
	}

	info("Usage: %s [-r <name=value>] <cgroup_path> ...\n", program_name);
	info("   or: %s --copy-from <source_cgroup_path> <cgroup_path> ...\n", program_name);
	info("Set the parameters of given cgroup(s)\n");
	info("  -1, --v1                                Provided parameters are in v1 format\n");
	info("  -2, --v2                                Provided parameters are in v2 format\n");
	info("  -i, --ignore-unmappable                 Do not return an error for ");
	info("settings that cannot be converted\n");
	info("  -r, --variable <name>                   Define parameter to set\n");
	info("  --copy-from <source_cgroup_path>        Control group whose ");
	info("parameters will be copied\n");
#ifdef WITH_SYSTEMD
	info("  -b                                      Ignore default systemd ");
	info("delegate hierarchy\n");
#endif
}
#endif /* !UNIT_TEST */

STATIC int parse_r_flag(const char * const program_name, const char * const name_value_str,
			struct control_value * const name_value)
{
	char *copy = NULL, *buf = NULL;
	int ret = 0;

	buf = strchr(name_value_str, '=');
	if (buf == NULL) {
		err("%s: wrong parameter of option -r: %s\n", program_name, optarg);
		ret = EXIT_BADARGS;
		goto err;
	}

	copy = strdup(name_value_str);
	if (copy == NULL) {
		err("%s: not enough memory\n", program_name);
		ret = -1;
		goto err;
	}

	/* parse optarg value */
	buf = strtok(copy, "=");
	if (buf == NULL) {
		err("%s: wrong parameter of option -r: %s\n", program_name, optarg);
		ret = EXIT_BADARGS;
		goto err;
	}

	strncpy(name_value->name, buf, FILENAME_MAX);
	name_value->name[FILENAME_MAX-1] = '\0';

	buf = strchr(name_value_str, '=');
	/*
	 * we don't need to check the return value of strchr because we
	 * know there's already an '=' character in the string.
	 */
	buf++;

	if (strlen(buf) == 0) {
		err("%s: wrong parameter of option -r: %s\n", program_name, optarg);
		ret = EXIT_BADARGS;
		goto err;
	}

	strncpy(name_value->value, buf, CG_CONTROL_VALUE_MAX);
	name_value->value[CG_CONTROL_VALUE_MAX-1] = '\0';

err:
	if (copy)
		free(copy);

	return ret;
}

#ifndef UNIT_TEST
int main(int argc, char *argv[])
{
	int ignore_default_systemd_delegate_slice = 0;
	struct control_value *name_value = NULL;
	int nv_number = 0;
	int nv_max = 0;

	struct cgroup *converted_src_cgroup = NULL;
	char src_cg_path[FILENAME_MAX] = "\0";
	struct cgroup *src_cgroup = NULL;
	struct cgroup *cgroup = NULL;

	enum cg_version_t src_version = CGROUP_UNK;
	bool ignore_unmappable = false;
	int ret = 0;
	int c;

	/* no parametr on input */
	if (argc < 2) {
		err("Usage is %s -r <name=value> <relative path to cgroup>\n", argv[0]);
		return -1;
	}

#ifdef WITH_SYSTEMD
	/* parse arguments */
	while ((c = getopt_long (argc, argv, "r:h12ib", long_options, NULL)) != -1) {
		switch (c) {
		case 'b':
			ignore_default_systemd_delegate_slice = 1;
			break;
#else
	while ((c = getopt_long (argc, argv, "r:h12i", long_options, NULL)) != -1) {
		switch (c) {
#endif
		case 'h':
			usage(0, argv[0]);
			ret = 0;
			goto err;
		case 'r':
			if ((flags &  FL_COPY) != 0) {
				usage(1, argv[0]);
				ret = EXIT_BADARGS;
				goto err;
			}
			flags |= FL_RULES;

			/* add name-value pair to buffer (= name_value variable) */
			if (nv_number >= nv_max) {
				nv_max += CG_NV_MAX;
				name_value = (struct control_value *)
					realloc(name_value, nv_max * sizeof(struct control_value));
				if (!name_value) {
					err("%s: not enough memory\n", argv[0]);
					ret = -1;
					goto err;
				}
			}

			ret = parse_r_flag(argv[0], optarg, &name_value[nv_number]);
			if (ret)
				goto err;

			nv_number++;
			break;
		case COPY_FROM_OPTION:
			if (flags != 0) {
				usage(1, argv[0]);
				ret = EXIT_BADARGS;
				goto err;
			}
			flags |= FL_COPY;
			strncpy(src_cg_path, optarg, FILENAME_MAX);
			src_cg_path[FILENAME_MAX-1] = '\0';
			break;
		case '1':
			src_version = CGROUP_V1;
			break;
		case '2':
			src_version = CGROUP_V2;
			break;
		case 'i':
			ignore_unmappable = true;
			break;
		default:
			usage(1, argv[0]);
			ret = EXIT_BADARGS;
			goto err;
		}
	}

	/* no cgroup name */
	if (!argv[optind]) {
		err("%s: no cgroup specified\n", argv[0]);
		ret = EXIT_BADARGS;
		goto err;
	}

	if (flags == 0) {
		err("%s: no name-value pair was set\n", argv[0]);
		ret = EXIT_BADARGS;
		goto err;
	}

	/* initialize libcgroup */
	ret = cgroup_init();
	if (ret) {
		err("%s: libcgroup initialization failed: %s\n", argv[0],
		    cgroup_strerror(ret));
		goto err;
	}

	/* this is false always for disable-systemd */
	if (!ignore_default_systemd_delegate_slice)
		cgroup_set_default_systemd_cgroup();

	/* copy the name-value pairs from -r options */
	if ((flags & FL_RULES) != 0) {
		src_cgroup = create_cgroup_from_name_value_pairs("tmp", name_value, nv_number);
		if (src_cgroup == NULL)
			goto err;
	}

	/* copy the name-value from the given group */
	if ((flags & FL_COPY) != 0) {
		src_cgroup = copy_name_value_from_cgroup(src_cg_path);
		if (src_cgroup == NULL)
			goto err;
	}

	while (optind < argc) {
		/* create new cgroup */
		cgroup = cgroup_new_cgroup(argv[optind]);
		if (!cgroup) {
			ret = ECGFAIL;
			err("%s: can't add new cgroup: %s\n", argv[0], cgroup_strerror(ret));
			goto err;
		}

		/* copy the values from the source cgroup to new one */
		ret = cgroup_copy_cgroup(cgroup, src_cgroup);
		if (ret != 0) {
			err("%s: cgroup %s error: %s\n", argv[0], src_cg_path,
			    cgroup_strerror(ret));
			goto err;
		}

		converted_src_cgroup = cgroup_new_cgroup(cgroup->name);
		if (converted_src_cgroup == NULL) {
			ret = ECGCONTROLLERCREATEFAILED;
			goto err;
		}

		/*
		 * If a failure occurs between this comment and the "End of above comment"
		 * below, then we need to manually free converted_src_cgroup.
		 */

		ret = cgroup_convert_cgroup(converted_src_cgroup, CGROUP_DISK, src_cgroup,
					    src_version);
		if ((ret && ret != ECGNOVERSIONCONVERT) ||
		    (ret == ECGNOVERSIONCONVERT && !ignore_unmappable)) {
			/*
			 * If the user not has specified that we ignore any errors
			 * due to being unable to map from v1 to v2 or vice versa,
			 * return error, else ignore the error and continue.
			 */
			free(converted_src_cgroup);
			goto err;
		}

		/*
		 * End of above comment about converted_src_cgroup needing to be manually freed.
		 */

		cgroup_free(&cgroup);
		cgroup = converted_src_cgroup;

		/* modify cgroup based on values of the new one */
		ret = cgroup_modify_cgroup(cgroup);
		if (ret) {
			err("%s: cgroup modify error: %s\n", argv[0], cgroup_strerror(ret));
			goto err;
		}

		optind++;
		cgroup_free(&cgroup);
	}

err:
	if (cgroup)
		cgroup_free(&cgroup);
	if (src_cgroup)
		cgroup_free(&src_cgroup);
	free(name_value);

	return ret;
}
#endif /* !UNIT_TEST */

#ifdef LIBCG_LIB
int cgroup_cgxset(const struct cgroup * const cgroup, enum cg_version_t version,
		  bool ignore_unmappable)
{
	struct cgroup *converted_cgroup;
	int ret;

	converted_cgroup = cgroup_new_cgroup(cgroup->name);
	if (converted_cgroup == NULL) {
		ret = ECGCONTROLLERCREATEFAILED;
		goto err;
	}

	ret = cgroup_convert_cgroup(converted_cgroup, CGROUP_DISK, cgroup, version);
	if (ret == ECGNOVERSIONCONVERT && ignore_unmappable)
		/*
		 * The user has specified that we should ignore any errors
		 * due to being unable to map from v1 to v2 or vice versa
		 */
		ret = 0;
	else if (ret)
		goto err;

	/* modify cgroup based on values of the new one */
	ret = cgroup_modify_cgroup(converted_cgroup);
	if (ret)
		goto err;

err:
	if (converted_cgroup)
		cgroup_free(&converted_cgroup);

	return ret;
}
#endif /* LIBCG_LIB */
